### ISSUE TITLE
Remove boringssl test

### ISSUE THREAD

**Issue by t8m** [openssl/openssl#14424]
See the failed run here:
https://github.com/openssl/openssl/pull/14416/checks?check_run_id=2030501177

```
key_agreement.go:19:2: code in directory /home/runner/work/openssl/openssl/boringssl/ssl/test/runner/curve25519 expects import "golang.org/x/crypto/curve25519"
chacha20_poly1305.go:23:2: code in directory /home/runner/work/openssl/openssl/boringssl/ssl/test/runner/poly1305 expects import "golang.org/x/crypto/poly1305"
../../../../util/wrap.pl go test -shim-path ../../../../test/ossl_shim/ossl_shim -shim-config ../../../../test/ossl_shim/ossl_config.json -pipe -allow-unimplemented => 1
not ok 1 - running BoringSSL tests
# ------------------------------------------------------------------------------
#   Failed test 'running BoringSSL tests'
#   at test/recipes/95-test_external_boringssl.t line 24.
# Looks like you failed 1 test of 1.95-test_external_boringssl.t .... 
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 
```

**Comment by t8m** [comment]
@davidben could perhaps give some hints?

**Comment by davidben** [comment]
tl;dr: Update to a later revision of BoringSSL.

ssl/test/runner/{poly1305,curve2551} were vendored copies of golang.org/x/crypto/{poly1305,curve25519} and imported via relative imports. It's a whole bunch of silliness from the days before [Go modules](https://blog.golang.org/using-go-modules). Looks like the immediate cause is that the vendored copies still had the [canonical import path](https://golang.org/doc/go1.4#canonicalimports) markers. I gather at some point Go started enforcing that on relative imports too?

Anyway, all this is unnecessary now and we got switched to Go modules [last year](https://boringssl-review.googlesource.com/c/boringssl/+/42044). So if you update to a revision past that point, I expect that'll be resolved.

**Comment by kaduk** [comment]
I will kick off a build with the boringssl commit that davidben linked, to see what happens...

**Comment by kaduk** [comment]
To give a status update on what I've been doing, in addition to the newer boringssl we also need to be using at least go 1.14.
The machine I've been testing this on is usually ubuntu bionic, but I seem to be able to install the go 1.14 packages from the ubuntu focal repository and force the boringssl test to use that go binary, which gets far enough that the boringssl test suite's handshake runner starts running feature queries of the test shim that the `ossl_shim` doesn't implement yet.  So, progress, but not clear yet how much work remains.

**Comment by kaduk** [comment]
With the following diff I'm able to get useful output from the boringssl tests again ... but there are 185 failing tests (!) and many will require some actual investigation to see if we are behaving dangerously.

```diff
diff --git a/boringssl b/boringssl
index 2070f8ad91..70fee17204 160000
--- a/boringssl
+++ b/boringssl
@@ -1 +1 @@
-Subproject commit 2070f8ad9151dc8f3a73bffaa146b5e6937a583f
+Subproject commit 70fee172042c9727f98ea2e6659eb77336e0dc7f
diff --git a/test/ossl_shim/ossl_config.json b/test/ossl_shim/ossl_config.json
index 1e57499065..48c5cf8b03 100644
--- a/test/ossl_shim/ossl_config.json
+++ b/test/ossl_shim/ossl_config.json
@@ -296,6 +296,18 @@
         ":BAD_CHANGE_CIPHER_SPEC:":"bad change cipher spec",
         ":ECC_CERT_NOT_FOR_SIGNING:":"ecc cert not for signing",
         ":OLD_SESSION_CIPHER_NOT_RETURNED:":"old session cipher not returned",
-        ":RESUMED_NON_EMS_SESSION_WITH_EMS_EXTENSION:":"inconsistent extms"
+        ":RESUMED_NON_EMS_SESSION_WITH_EMS_EXTENSION:":"inconsistent extms",
+        ":NO_SUPPORTED_VERSIONS_ENABLED:":"protocol version not supported",
+        ":APPLICATION_DATA_INSTEAD_OF_HANDSHAKE:":"unexpected message",
+        ":UNKNOWN_ALERT_TYPE:":"illegal parameter",
+        ":BAD_HELLO_REQUEST:":"illegal parameter",
+        ":UNSUPPORTED_COMPRESSION_ALGORITHM:":"illegal parameter",
+        ":UNEXPECTED_EXTENSION:":"unsupported extension",
+        ":DECODE_ERROR:":"error decoding message",
+        ":NO_COMMON_SIGNATURE_ALGORITHMS:":"handshake failure",
+        ":BAD_ECPOINT:":"no cipher suite supported by both client and server",
+        ":EXCESS_HANDSHAKE_DATA:":"unexpected message",
+        ":CANNOT_PARSE_LEAF_CERT:":"bad certificate",
+        ":KEY_USAGE_BIT_INCORRECT:":"handshake failure"
     }
 }
diff --git a/test/ossl_shim/ossl_shim.cc b/test/ossl_shim/ossl_shim.cc
index 380e6853c6..0aea07b3fd 100644
--- a/test/ossl_shim/ossl_shim.cc
+++ b/test/ossl_shim/ossl_shim.cc
@@ -487,7 +487,7 @@ static int ServerNameCallback(SSL *ssl, int *out_alert, void *arg) {
 // Connect returns a new socket connected to localhost on |port| or -1 on
 // error.
 static int Connect(uint16_t port) {
-  int sock = socket(AF_INET, SOCK_STREAM, 0);
+  int sock = socket(AF_INET6, SOCK_STREAM, 0);
   if (sock == -1) {
     PrintSocketError("socket");
     return -1;
@@ -499,11 +499,11 @@ static int Connect(uint16_t port) {
     closesocket(sock);
     return -1;
   }
-  sockaddr_in sin;
+  sockaddr_in6 sin;
   memset(&sin, 0, sizeof(sin));
-  sin.sin_family = AF_INET;
-  sin.sin_port = htons(port);
-  if (!inet_pton(AF_INET, "127.0.0.1", &sin.sin_addr)) {
+  sin.sin6_family = AF_INET6;
+  sin.sin6_port = htons(port);
+  if (!inet_pton(AF_INET6, "::1", &sin.sin6_addr)) {
     PrintSocketError("inet_pton");
     closesocket(sock);
     return -1;
@@ -1270,6 +1270,12 @@ static int Main(int argc, char **argv) {
   signal(SIGPIPE, SIG_IGN);
 #endif
 
+  // handle queries from the test framework
+  if (argc == 2 && strcmp(argv[1], "-is-handshaker-supported") == 0) {
+    fprintf(stdout, "No\n");
+    return 0;
+  }
+
   OPENSSL_init_crypto(0, NULL);
   OPENSSL_init_ssl(0, NULL);
   g_config_index = SSL_get_ex_new_index(0, NULL, NULL, NULL, NULL);
diff --git a/test/ossl_shim/test_config.cc b/test/ossl_shim/test_config.cc
index a37d010d7a..8a99b8d29b 100644
--- a/test/ossl_shim/test_config.cc
+++ b/test/ossl_shim/test_config.cc
@@ -104,6 +104,7 @@ const Flag<std::string> kStringFlags[] = {
   { "-cipher", &TestConfig::cipher },
   { "-export-label", &TestConfig::export_label },
   { "-export-context", &TestConfig::export_context },
+  { "-handshaker-path", &TestConfig::handshaker_path },
 };
 
 const Flag<std::string> kBase64Flags[] = {
diff --git a/test/ossl_shim/test_config.h b/test/ossl_shim/test_config.h
index 6dc7319485..f6d17cc422 100644
--- a/test/ossl_shim/test_config.h
+++ b/test/ossl_shim/test_config.h
@@ -59,6 +59,7 @@ struct TestConfig {
   int export_keying_material = 0;
   std::string export_label;
   std::string export_context;
+  std::string handshaker_path;
   bool use_export_context = false;
   bool expect_ticket_renewal = false;
   bool expect_no_session = false;
diff --git a/test/recipes/95-test_external_boringssl.t b/test/recipes/95-test_external_boringssl.t
index ea607935aa..763c9d4564 100644
--- a/test/recipes/95-test_external_boringssl.t
+++ b/test/recipes/95-test_external_boringssl.t
@@ -21,7 +21,7 @@ plan skip_all => "BoringSSL runner not detected"
 plan tests => 1;
 
 indir $ENV{BORING_RUNNER_DIR} => sub {
-    ok(run(cmd(["go", "test", "-shim-path",
+    ok(run(cmd(["/usr/lib/go-1.14/bin/go", "test", "-shim-path",
                 bldtop_file("test", "ossl_shim", "ossl_shim"),
                 "-shim-config",
                 srctop_file("test", "ossl_shim", "ossl_config.json"),
```

(I have several small independent commits to make these changes that have a bit more commentary.  Hardcoding `/usr/lib/go-1.14/bin/go` is obviously only good for local experimentation.)

**Comment by kaduk** [comment]
The full (verbose) log output:

[t1.log](https://github.com/openssl/openssl/files/6160290/t1.log)

**Comment by t8m** [comment]
Could we use the
```
sudo add-apt-repository ppa:longsleep/golang-backports
sudo apt update
sudo apt install golang-go
```
to install golang?
Not sure it is good idea to depend on ppa though.
And we should skip the test if go-1.14 or newer is not available.

**Comment by mattcaswell** [comment]
> but there are 185 failing tests (!)

This was what I was afraid of. I have a feeling that the previous (working) boringssl tests were pre-TLSv1.3...or pre-final-TLSv1.3. I guess there have been many tests added since then and boringssl and openssl have probably diverged even more in the meantime.

I am wondering whether the boringssl external tests are worth the maintenance burden.

**Comment by t8m** [comment]
We should discuss the potential removal within OTC.

**Comment by kaduk** [comment]
> > but there are 185 failing tests (!)
> 
> This was what I was afraid of. I have a feeling that the previous (working) boringssl tests were pre-TLSv1.3...or pre-final-TLSv1.3. I guess there have been many tests added since then and boringssl and openssl have probably diverged even more in the meantime.
> 
> I am wondering whether the boringssl external tests are worth the maintenance burden.

FWIW, my own opinion is that the boringssl external tests are worth the maintenance burden; they have caught several things for us in the past.  Having skimmed through the test log (before posting), there are clearly not 185 unique issues: e.g., there's a cluster of 10-20 relating to TLS 1.3 vs 1.2 as an expected version, another cluster involving messages received after shutdown (which AFAIK we don't even claim to try to implement anything for), etc.

**Comment by t8m** [comment]
On the other hand the failures are there most probably for 1.1.1 as well and it will take definitely non-trivial amount of work to fix them. Should this be a 3.0-beta1 blocker work?

**Comment by kaduk** [comment]
> On the other hand the failures are there most probably for 1.1.1 as well and it will take definitely non-trivial amount of work to fix them. Should this be a 3.0-beta1 blocker work?

From my point of view what would have been a beta1 blocker is getting the same tests that we "used to" run in a passing state.  With a newer boringssl bringing in additional tests, that additional work need not be scoped to beta1.

I think that even if we can't get the boringssl tests integrated into the CI we should still run the current state of them manually and triage at the results (if not passing) before releasing beta1.

**Comment by davidben** [comment]
I expect it'll always be a bit imperfect just because OpenSSL and BoringSSL support different features, but I'm happy to explore ways to make that smoother, while still testing what we need to test on our end. (Ben recently needing to ignore a silly `-handshaker-path` flag, so I have a change out to avoid that for you all.)

We make a lot of combinatorial tests, so the raw numbers may be a little misleading. For instance, the `CipherNegotiation-*` failures seem to be because OpenSSL doesn't support our equipreference syntax for cipher suites. (We also made the cipher strings accept the IETF names, which I'm not sure if OpenSSL supports.) The DisabledTests field supports wildcard patterns, so you can simply add "CipherNegotiation-*" to there.

The error mapping appears to be another tricky point. The problem is OpenSSL, and thus BoringSSL, have absurdly specific error codes, which makes it tricky to translate tests from one implementation to another. I suspect we want to use the shim_config.json features to lower the resolution on both runner and shim side when running with OpenSSL. Whether the ErrorMap mechanism is sufficient or if we want something else (test-specific error maps?), I'm not sure.

(Incidentally, it looks like your shim config still disables all TLS 1.3 tests, so there're probably some tests to look at enabling after that.)

**Comment by mattcaswell** [comment]
OTC: We will drop the boringssl tests.

**Comment by briansmith** [comment]
> OTC: We will drop the boringssl tests.

I don't know the reasoning but this seems like a bad idea and an overreaction to me. Rustls, NSS, and other implementations that are even more different from BoringSSL than OpenSSL have found ways to use the BoringSSL tests and the BoringSSL test suite is effectively the standard test suite for TLS at this point.
